### PR TITLE
Enable page-locked tensors without CUDA

### PR DESCRIPTION
**Pull request by tjruwase**
Enable CPU page-locked tensors in non-CUDA environments.

### CODE REVIEW

File: csrc/aio/py_lib/deepspeed_pin_tensor.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+Copyright 2023 The Microsoft DeepSpeed Team
+Licensed under the MIT license.
+
+Functionality for managing pinned CPU memory.
+*/
+
+#include <map>
+#include "deepspeed_py_aio.h"
+
+struct deepspeed_pin_tensor_t {
```
[the remark this answers is not in the document]
Comment (ShadenSmith) [reply csrc/aio/py_lib/deepspeed_pin_tensor.h:11]: (ultimately, just thinking how to make it harder to leak pinned memory allocations)

File: csrc/aio/py_lib/deepspeed_pin_tensor.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright 2023 The Microsoft DeepSpeed Team
+Licensed under the MIT license.
+
+Functionality for managing pinned CPU memory.
+*/
+
+#include <map>
```
[the remark this answers is not in the document]
Comment (tjruwase) [reply csrc/aio/py_lib/deepspeed_pin_tensor.h:11]: Fixed.

File: csrc/aio/py_lib/deepspeed_pin_tensor.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjruwase) [reply csrc/aio/py_lib/deepspeed_pin_tensor.h:14]: Yes, your instincts are spot on. This is work in progress, and I should have marked code as such. Sorry for the confusion. The final goal is a page-locked tensor manager. I decided to start with a basic implementation just to unblock testing. 

Note that the current code matches the semantics of existing CUDA pinned memory semantics, i.e., no unpinning logic.

File: csrc/aio/py_lib/deepspeed_py_aio_handle.h [hunk 3]
```diff
@@ -56,6 +58,10 @@ struct deepspeed_aio_handle_t {
 
     int async_pwrite(const torch::Tensor& buffer, const char* filename);
 
+    torch::Tensor new_cpu_locked_tensor(const size_t num_elem, const torch::Tensor& example_tensor);
```
[the remark this answers is not in the document]
Comment (tjruwase) [reply csrc/aio/py_lib/deepspeed_py_aio_handle.h:62]: I would love to chat more about this. I could not find a direct way to pass `dtype` and was further discouraged by this [post](https://discuss.pytorch.org/t/how-to-pass-python-device-and-dtype-to-c/69577/).

I would mark this as WIP as well.

[the remark this answers is not in the document]
Comment (ShadenSmith) [reply csrc/aio/py_lib/deepspeed_py_aio_handle.h:62]: Oh interesting! That's news to me, in that case never mind 😃

File: csrc/aio/py_lib/deepspeed_pin_tensor.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShadenSmith) [reply csrc/aio/py_lib/deepspeed_pin_tensor.h:14]: Got it, that makes sense thanks!! Excited for this work :-)

**Review by mrwyattii — APPROVED** [review]
LGTM

File: csrc/aio/py_lib/deepspeed_py_aio_handle.cpp [hunk 4]
```diff
@@ -280,3 +281,14 @@ int deepspeed_aio_handle_t::async_pwrite(const torch::Tensor& buffer, const char
 {
     return pwrite(buffer, filename, false, true);
 }
+
```
Comment (xiexbing) [inline csrc/aio/py_lib/deepspeed_py_aio_handle.cpp:284]: so, here multithreading is used to enable aio writes, where block size represents dirty pages, submit, overlap are system calls, queue is used to manage the pages?

Comment (tjruwase) [reply csrc/aio/py_lib/deepspeed_py_aio_handle.cpp:284]: Partially correct. Let's chat more offline.

**Comment by tjruwase** [comment]
Bypassing AMD CI

**Pull request merged**